storey/sources.py

```diff
@@ -619,20 +620,33 @@ def _blocking_io_loop(self):
                                 for single_key_field in self._key_field:
                                     if self._with_header and isinstance(single_key_field, str):
                                         single_key_field = field_name_to_index[single_key_field]
+                                    if parsed_line[single_key_field] is None:
+                                        create_event = False
+                                        break
                                     key.append(parsed_line[single_key_field])
                             else:
                                 key_field = self._key_field
                                 if self._with_header and isinstance(key_field, str):
                                     key_field = field_name_to_index[key_field]
                                 key = parsed_line[key_field]
+                                if key is None:
+                                    create_event = False
                         if self._time_field:
                             time_field = self._time_field
                             if self._with_header and isinstance(time_field, str):
                                 time_field = field_name_to_index[time_field]
                             time_as_datetime = parsed_line[time_field]
                         else:
                             time_as_datetime = datetime.now()
-                        self._event_buffer.put(Event(element, key=key, time=time_as_datetime))
+                        event = Event(element, key=key, time=time_as_datetime)
+                        if create_event:
```

Comment:
Better move the work for creating the event into the if, to avoid unnecessary work.